storage_controller/src/service.rs

```diff
@@ -3547,6 +3557,309 @@ impl Service {
+    ) -> Result<Vec<NodeId>, ApiError> {
+        // If quorum is reached, return if we are outside of a specified timeout
+        let jwt = self
+            .config
+            .safekeeper_jwt_token
+            .clone()
+            .map(SecretString::from);
+        let mut joinset = JoinSet::new();
+
+        let safekeepers = {
+            let locked = self.inner.read().unwrap();
+            locked.safekeepers.clone()
+        };
+
+        let mut members = Vec::new();
+        for sk_id in timeline_persistence.sk_set.iter() {
+            let sk_id = NodeId(*sk_id as u64);
+            let Some(safekeeper) = safekeepers.get(&sk_id) else {
+                return Err(ApiError::InternalServerError(anyhow::anyhow!(
+                    "couldn't find entry for safekeeper with id {sk_id}"
+                )))?;
+            };
+            members.push(SafekeeperId {
+                id: sk_id,
+                host: safekeeper.skp.host.clone(),
+                pg_port: safekeeper.skp.port as u16,
+            });
+        }
+        let mut mconf = safekeeper_api::membership::Configuration::empty();
+        mconf.members = MemberSet::new(members).map_err(ApiError::InternalServerError)?;
+
+        let req = safekeeper_api::models::TimelineCreateRequest {
+            commit_lsn: None,
+            mconf,
+            pg_version,
+            start_lsn: timeline_persistence.start_lsn.0,
+            system_id: None,
+            tenant_id,
+            timeline_id,
+            wal_seg_size: None,
+        };
+        const SK_CREATE_TIMELINE_RECONCILE_TIMEOUT: Duration = Duration::from_secs(30);
+        for sk in timeline_persistence.sk_set.iter() {
+            let sk_id = NodeId(*sk as u64);
+            let safekeepers = safekeepers.clone();
+            let jwt = jwt.clone();
+            let req = req.clone();
+            // Unwrap is fine as we already would have returned error above
+            joinset.spawn(async move {
+                let sk_p = safekeepers.get(&sk_id).unwrap();
+                sk_p.with_client_retries(
+                    |client| {
+                        let req = req.clone();
+                        async move { client.create_timeline(&req).await }
+                    },
+                    &jwt,
+                    3,
+                    3,
+                    SK_CREATE_TIMELINE_RECONCILE_TIMEOUT,
+                    &CancellationToken::new(),
+                )
+                .await
+                .map_err(|e| (sk_id, e))?;
+                Ok(())
+            });
+        }
+        // After we have built the joinset, we now wait for the tasks to complete,
+        // but with a specified timeout to make sure we return swiftly, either with
+        // a failure or success.
+        let reconcile_deadline = tokio::time::Instant::now() + SK_CREATE_TIMELINE_RECONCILE_TIMEOUT;
+
+        // Treat the first two tasks to finish differently, mostly when they timeout,
```

Comment:
Let's avoid hardcoding 2/3 and do proper majorities. Moreover, seems like next 50 lines would sensibly simpler if coded like
```
loop {
  select! {
     joinset.join_next() => if Some(res) { log and remember result } else break;
     timeout => break;
  } 
}
// Look at all collected results, return Ok if have good majority, Err otherwise, better including in err msg if timeout reached.
```

In any case it's important to log all results.

Reply:
using a loop now

Reply:
It can be even simpler. 
- counting `outstanding` is not needed, we can just check quorum once in the end, when all tasks are joined or timeout happened. Plus it would avoid unwrap. timeout can be logged separately immediately when it is hit, or appended to later logging.
- again, let's avoid hardcoding quorums. The idea is to have NUM_SAFEKEEPERS = 3 const so that we can e.g. change it to 5 or 1 and nothing else needs to be adjusted. It's eventually should be a timeline creation / storcon param as we often use 1 sk in tests.
- let's log each result, like info!("successfully created timeline on safekeeper {}") or failed to do so.

It may be annoying, but things like that do huge difference when investigating what's happening in the field.

Reply:
simplified it even further. regarding the quorums question, I suppose we should figure out a way how to configure the storage controller for single-sk operations. probably it should default to three safekeepers by default, but there is some flag or something to enable single-sk mode.